addition of S1 (componenets (#96)
OmarHassanAdelhamid/Five-of-a-Kind-capstone-project-: docs/SRS-Meyer/requirements/parts/system/S2.adoc

[#s2,reftext=S.2]
=== (S.2) Functionality

ifdef::!env-draft[]
TIP: _**This is the bulk of the System book, describing elements of functionality (behaviors)**. This chapter corresponds to the traditional view of requirements as defining "**what the system does**”. It is organized as one section, S.2.n, for each of the components identified in <<s1>>, describing the corresponding behaviors (functional and non-functional properties)._  <<BM22>>
endif::[]

==== (2.2.1) Slicing Manager

===== Functional Requirements

. [[F211]] *Define voxel block size:* Slicing Manager shall allow the user to configure the dimensions of the rectangular voxel prisms used to slice their geometric model. (F211)

> . [[F212]] *Adjust geometric model size:* Slicing Manager shall allow users to modify the dimensions of their model while verifying that it is evenly divisible by the chosen voxel size. (F212)

Reviewer: will need to check with onaizah how to deal with whether it's evenly divisible or not. also, is this stating that the cad file can be modified, or the total dimensions of the voxel area? may need to reword for clarity

> . [[F213]] *Start new project:* Slicing Manager shall allow a user to import a CAS from their personal device in order to start a new project. (F213)

Reviewer: ```suggestion
. [[F213]] *Start new project:* Slicing Manager shall allow a user to import a CAD file from their personal device in order to start a new project. (F213)
```

<<<
===== Non-Functional Requirements

> . [[NF211]] *Scalable projects:* Slicing Manager shall support the generation of sliced models that contain up to 10,000 voxels without loss of functionality.  (NF211)

Reviewer: cross-check with info Onaizah gave us in terms of numbers (ignore if this is correct)

. [[NF212]] *Slicing process performance:* Slicing Manager shall generate the voxel data file at a minimum rate of 1000 voxels per second. (NF212)

==== (2.2.2) Imaging Manager

===== Functional Requirements

> . [[F221]] *Generate 3D image:* Imaging Manager shall recreate a 3D modelling image with clear visualization of sliced voxels. (F221)

Reviewer: I don't think recreate is the right word, maybe generate image instead? (drop modelling)

> . [[F222]] *Adjust image:* Imaging Manager shall provide users with an intuitive interface that permits seamless navigation across multiple perspectives of the 3D model through rotation and resizing. (F222)

Reviewer: drop seamless, could be considered vague.

. [[F223]] *Highlight selection:* Imaging manager shall provide users with visualization showcasing what part of the 3D model is currently selected. (F223)

> . [[F224]] *Recreate previous project:* Slicing Manager shall be able to accurately interpret and recreate a previously saved geometric model that has already been sliced and magnetized. (F224)

Reviewer: why is a slicing manager requirement listed under imaging manager? also, may want to rephrase interpret/recreate to load, and not mention magnetized.

===== Non-Functional Requirements

. [[NF221]] *Image updates:* Imaging Manager shall update any changes to the perspective of the 3D model with a latency of less than 500ms to allow a seamless user interaction. (NF221)

. [[NF222]] *Visual scalability:* Imaging Manager shall support the visual display of voxel models that contain up to 10,000 voxels without significant degradation of performance or loss of functionality. (NF222)

==== (2.2.3) Magnetization Manager

===== Functional Requirements

> . [[F231]] *View layer:* Magnetization manager shall provide users with an interface to choose a layer of their model, in which they can then easily view and select certain groups of voxels. (F231)

Reviewer: Khalid also mentioned being able to edit multiple layers at once; we need to decide if we are doing this or not (inconsistency, should bring this up with the group)

> . [[F232]] *Set magnetization:* Magnetization manager shall allow users to set a specific magnetization value for either an individual or a  group of selected voxels. (F232)

Reviewer: khalid also mentions being able to use pre-set vectors or custom ones; need to decide if we're doing this or not.

. [[F233]] *Magnetization tracker:* Magnetization manager shall integrate easy tracking of what has already been magnetized by adjusting the colour of the voxel. (F233)

> . [[F234]] *Favourite bar:* Magnetization manager shall allow users to define and maintain a list of ‘favourite’ magnetizations for quick selection and easy reuse. (F234)

Reviewer: this is new functionality ;  will need to cross-check other sections to ensure consistency.

===== Non-Functional Requirements

. [[NF231]] *Ease of use:* Magnetization manager shall ensure that the process of magnetizing a large group of selected voxels can be completed in 5 steps or less, creating an intuitive interface. (NF231)

. [[NF232]] *Metadata update:* Magnetization manager shall update the voxel magnetization metadata with a latency of less than 1s to allow consistent, synchronized model modification updates. (NF232)

==== (2.2.4) Exportation Manager

===== Functional Requirements

> . [[F241]] *Magnetization validation:* Exportation Manager shall validate that all voxels have received a magnetization value upon receiving a request to export a file for printing. (F241)

Reviewer: we should check with onaizah if it's possible for voxels to not be magnetized, eg. if we should put null values. (unlikely, but still)

> . [[F242]] *Add to recent projects:* Exportation Manager shall add the file to a list of recent projects visible to the user for quick access and future edits. (F242)

Author: Decides to get ride of that requirement and replace it with something I think is more useful (I was having difficulties coming up with some so honestly I don't feel like that requirement was that good)

> . [[F243]] *Export to print:* Exportation Manager shall generate a file that contains all metadata for each voxel in a structured format. (F243)

Reviewer: we also need the csv option as well. (eg. an additional file that is exported.)

===== Non-Functional Requirements:

> . [[NF241]] *Printing compliance:* Exportation Manager shall ensure that the exported file is compatible and easily interpreted by the existing Java software that is used to interpret geometric models for printing. (NF241)

Reviewer: ```suggestion
. [[NF241]] *Printing compliance:* Exportation Manager shall ensure that the exported file is compatible with the existing Java software that is used to interpret geometric models for printing. (NF241)
```

> . [[NF241]] *Printing compliance:* Exportation Manager shall ensure that the exported file is compatible and easily interpreted by the existing Java software that is used to interpret geometric models for printing. (NF241)

Reviewer: I'm not sure if we need to formalize this again as a requirement after I've listed it as a constraint in one of the E sections
